transformations/transformations/scc_annotate.py

```diff
@@ -0,0 +1,323 @@
+                # Skip loops explicitly marked with `!$loki/claw nodep`
+                if loop.pragma and any('nodep' in p.content.lower() for p in as_tuple(loop.pragma)):
+                    continue
+
+                if loop.variable != horizontal.index:
+                    # Perform pragma addition in place to avoid nested loop replacements
+                    loop._update(pragma=(ir.Pragma(keyword='acc', content='loop seq'),))
+
+    @classmethod
+    def kernel_annotate_subroutine_present_openacc(cls, routine):
+        """
+        Insert ``!$acc data present`` annotations around the body of a subroutine.
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            The subroutine to which annotations will be added
+        """
+
+        # Get the names of all array and derived type arguments
+        args = [a for a in routine.arguments if isinstance(a, sym.Array)]
+        args += [a for a in routine.arguments if isinstance(a.type.dtype, DerivedType)]
+        argnames = [str(a.name) for a in args]
+
+        routine.body.prepend(ir.Pragma(keyword='acc', content=f'data present({", ".join(argnames)})'))
+        # Add comment to prevent false-attachment in case it is preceded by an "END DO" statement
+        routine.body.append((ir.Comment(text=''), ir.Pragma(keyword='acc', content='end data')))
+
+    @classmethod
+    def insert_annotations(cls, routine, horizontal, vertical, hoist_column_arrays):
+
+        # Mark all non-parallel loops as `!$acc loop seq`
+        cls.kernel_annotate_sequential_loops_openacc(routine, horizontal)
+
+        # Mark all parallel vector loops as `!$acc loop vector`
+        cls.kernel_annotate_vector_loops_openacc(routine, horizontal, vertical)
+
+        # Wrap the routine body in `!$acc data present` markers
+        # to ensure device-resident data is used for array and struct arguments.
+        cls.kernel_annotate_subroutine_present_openacc(routine)
+
+        if hoist_column_arrays:
+            # Mark routine as `!$acc routine seq` to make it device-callable
+            routine.spec.append(ir.Pragma(keyword='acc', content='routine seq'))
+
+        else:
+            # Mark routine as `!$acc routine vector` to make it device-callable
+            routine.spec.append(ir.Pragma(keyword='acc', content='routine vector'))
+
+    def transform_subroutine(self, routine, **kwargs):
+        """
+        Apply SCCAnnotate utilities to a :any:`Subroutine`.
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            Subroutine to apply this transformation to.
+        role : string
+            Role of the subroutine in the call tree; should be ``"kernel"``
+        """
+
+        # TODO: we only need this here until the scheduler can combine multiple transformations into single pass
+        # Bail if routine has already been processed
+        if self._processed.get(routine, None):
```

Comment:
[No action] Indeed, this mechanism isn't great, but will likely only become redundant once we can preempt repeated processing, which should come with the scheduler rewrites. Ok for now, I htink.

Reply:
The scheduler already ensures that each item is visited only once, so shouldn't this instead be replaced by the typical pattern?
```python
if (item := kwargs.get('item')) and item.name != routine.name.lower():
    return
```
I think this should become part of the dispatch in the `Transformation.apply` routine eventually.

Reply:
Yes, I agree. I was planning on moving these bail-out mechanism when I switch to composed SCC for full ec-physics, but we can bring them in line here already

Reply:
I think it's not necessary to move this already to the base class as part of this PR, but would rather not introduce a second pattern.

Reply:
Yes, fully agreed. I should have realised that this is just duplicating an existing mechanism. I'll replace all instances of the `self._processed` checks with B's suggestion above.